edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,812 @@
+            this.moduleIdentityLifecycleManager = moduleIdentityLifecycleManager;
+            this.client = client;
+        }
+
+        public async Task ListCrdComplete(Task<HttpOperationResponse<object>> customObjectWatchTask)
+        {
+            if (customObjectWatchTask != null)
+            {
+                HttpOperationResponse<object> customObjectWatch = await customObjectWatchTask;
+                if (customObjectWatch != null)
+                {
+                    // We can add events to a watch once created, like if connection is closed, etc.
+                    this.operatorWatch = Option.Some(
+                        customObjectWatch.Watch<object>(
+                            onEvent: async (type, item) =>
+                            {
+                                try
+                                {
+                                    await this.WatchDeploymentEventsAsync(type, item);
+                                }
+                                catch (Exception ex) when (!ex.IsFatal())
+                                {
+                                    this.logger.ExceptionInCustomResourceWatch(ex);
+                                }
+                            },
+                            onClosed: () =>
+                            {
+                                this.logger.CrdWatchClosed();
+
+                                // get rid of the current crd watch object since we got closed
+                                this.operatorWatch.ForEach(watch => watch.Dispose());
+                                this.operatorWatch = Option.None<Watcher<object>>();
+
+                                // kick off a new watch
+                                this.client.ListClusterCustomObjectWithHttpMessagesAsync(
+                                    Constants.K8sCrdGroup,
+                                    Constants.K8sApiVersion,
+                                    Constants.K8sCrdPlural,
+                                    watch: true).ContinueWith(this.ListCrdComplete);
+                            },
+                            onError: this.logger.ExceptionInCustomResourceWatch
+                        ));
+                }
+                else
+                {
+                    this.logger.NullListResponse("ListClusterCustomObjectWithHttpMessagesAsync", "http response");
+                    throw new NullReferenceException("Null response from ListClusterCustomObjectWithHttpMessagesAsync");
+                }
+            }
+            else
+            {
+                this.logger.NullListResponse("ListClusterCustomObjectWithHttpMessagesAsync", "task");
+                throw new NullReferenceException("Null Task from ListClusterCustomObjectWithHttpMessagesAsync");
+            }
+        }
+
+        private async Task WatchDeploymentEventsAsync(WatchEventType type, object custom)
+        {
+            using (await this.watchLock.LockAsync())
```

Comment:
>watchLock.LockAsync [](start = 30, length = 19)

This is a pretty large section of code to block over.

Reply:
I dropped the payload to only the part that operates.